mlir/lib/Dialect/Arith/IR/ArithOps.cpp

```diff
@@ -400,6 +400,35 @@ OpFoldResult arith::MulIOp::fold(FoldAdaptor adaptor) {
       [](const APInt &a, const APInt &b) { return a * b; });
 }
 
+void arith::MulIOp::getAsmResultNames(
+    function_ref<void(Value, StringRef)> setNameFn) {
+  if (!isa<IndexType>(getType()))
+    return;
+
+  // Match vector.vscale by name to avoid depending on the vector dialect (which
+  // is a circular dependency).
+  auto isVscale = [](Operation *op) {
+    return op && op->getName().getStringRef() == "vector.vscale";
+  };
+
+  // Name `base * vscale` or `vscale * base` as `c<base_value>_vscale`.
+  IntegerAttr baseValue;
+  if (matchPattern(getLhs(), m_Constant(&baseValue)) &&
+      isVscale(getRhs().getDefiningOp())) {
+    // base * vscale
+  } else if (isVscale(getLhs().getDefiningOp()) &&
+             matchPattern(getRhs(), m_Constant(&baseValue))) {
+    // vscale * base
+  } else {
+    return;
+  }
```

Comment:
Could we avoid `if/else-if` nests? How about:

```suggestion
  bool isVscaleXBase = matchPattern(getLhs(), m_Constant(&baseValue)) &&
      isVscale(getRhs().getDefiningOp());
  bool isBaseXVscale = isVscale(getLhs().getDefiningOp()) &&
             matchPattern(getRhs(), m_Constant(&baseValue));
  if (!(isVscaleXBase || isBaseXVscale)
    return;
```

Reply:
I personally find the checks much more readable when kept separate.

Reply:
> I personally find the checks much more readable when kept separate.

could move it to a lambda:
```suggestion
  auto isVscaleExpr = [&](Value a, Value b) {
    return matchPattern(a, m_Constant(&baseValue)) &&
        isVscale(b.getDefiningOp());
  };

  // Name `base * vscale` or `vscale * base` as `c<base_value>_vscale`.
  if (!isVscaleExpr(getLhs(), getRhs()) && !isVscaleExpr(getRhs(), getLhs()))
    return;
```

Reply:
> I personally find the checks much more readable when kept separate.

Hm, that's exactly what I wanted to emphasise :) To me having two separate bools to "self-document" what the condition is make things easier (no need for comments):
```
  bool isVscaleXBase; // instead of adding a comment: "vscale * base"
  bool isBaseXVscale; // instead of adding a comment: "base * vscale"
```

In particular, it keeps them "more" separate than having nested `else-if`, no? In any case, I just want to avoid `else if` :) Cullen's suggestion for a lambda works as well.

Reply:
`isBaseXVscale` is a fairly cryptic name to me (I would not get the meaning without reading the match expression). My brain just reads it as `Is base XV scale?`, then has to look at the code.

Reply:
Also, `else-if` clearer than an expression `!(a || b)`. I see an `else-if` and I know I'm looking at some code that's looking for one case (that's hardwired). I see multiple bools and an expression, I have to stop parse the expression and see what's happening.

Reply:
(general point here is self documenting code is a lie :stuck_out_tongue:, what's self documenting to you is not to me, and likely vise-versa)

Reply:
The lambda has the advantage over the two bool that it preserves the shortcut evaluation (we don't want to match both if one is enough)

> (general point here is self documenting code is a lie 😛, what's self documenting to you is not to me, and likely vise-versa)

This is true to some extent, but this is why we stick to idiom and keep the codebase consistent. It may not feel self-documenting at first, but this mostly a matter of habit.